File: src/api/reports.ts
const API_BASE_URL = 'http://localhost:8000/api/v1';

export interface ReportTemplateApiData {
  id: string;
  key: string;
  title: string;
  subtitle: string;
  description: string;
  category: 'Executive' | 'Technical' | 'Compliance';
  icon: string;
}

export interface ReportGeneratePayload {
  project_id: number;
  report_type: string;
  format: 'pdf' | 'html' | 'json';
}

export const fetchReportTemplates = async (): Promise<ReportTemplateApiData[]> => {
  const response = await fetch(`${API_BASE_URL}/reports/templates`);
  if (!response.ok) {
    throw new Error('Failed to fetch report templates');
  }
  return response.json();
};

export const generateReportJson = async (projectId: number, reportType: string): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/reports/generate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      project_id: projectId,
      report_type: reportType,
      format: 'json',
    }),
  });

  if (!response.ok) {
    throw new Error('Failed to generate report data');
  }
  return response.json();
};

export const generateReportHtml = async (projectId: number, reportType: string): Promise<string> => {
  const response = await fetch(`${API_BASE_URL}/reports/generate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      project_id: projectId,
      report_type: reportType,
      format: 'html',
    }),
  });

  if (!response.ok) {
    throw new Error('Failed to generate HTML report');
  }
  return response.text();
};

export const downloadReportPdf = async (projectId: number, reportType: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/reports/generate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      project_id: projectId,
      report_type: reportType,
      format: 'pdf',
    }),
  });

  if (!response.ok) {
    throw new Error('Failed to generate PDF report');
  }

  const blob = await response.blob();
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.style.display = 'none';
  a.href = url;
  a.download = `kyptic_report_proj${projectId}_${reportType}.pdf`;
  document.body.appendChild(a);
  a.click();
  window.URL.revokeObjectURL(url);
  document.body.removeChild(a);
};

File: backend/app/routers/reports.py
import html
import re
from typing import Any, Dict, List
from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from fastapi.responses import Response, HTMLResponse, JSONResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.services.report_service import ReportService


router = APIRouter(prefix="/reports", tags=["Reports"])


class ReportGenerateRequest(BaseModel):
    project_id: int
    report_type: str = "executive"  # executive, developer, compliance, owasp
    format: str = "pdf"             # pdf, html, json


@router.get("/templates")
def get_report_templates() -> List[Dict[str, Any]]:
    return [
        {
            "id": "t-1",
            "key": "executive",
            "title": "Executive Summary",
            "subtitle": "High-level risk overview & security score",
            "description": "Tailored for leadership and CISOs. Summarizes overarching security posture, critical risk score (0-100), and severity trends.",
            "category": "Executive",
            "icon": "assignment_ind",
        },
        {
            "id": "t-2",
            "key": "developer",
            "title": "Developer Report",
            "subtitle": "Technical deep-dive & code snippets",
            "description": "Granular technical details including stack traces, affected code snippets, line numbers, and actionable remediation steps.",
            "category": "Technical",
            "icon": "code",
        },
        {
            "id": "t-3",
            "key": "compliance",
            "title": "Compliance Report",
            "subtitle": "PCI-DSS & SOC2 regulatory mapping",
            "description": "Maps discovered vulnerabilities directly to specific regulatory clauses within frameworks like PCI-DSS v4.0 and SOC2 Type II.",
            "category": "Compliance",
            "icon": "verified_user",
        },
        {
            "id": "t-4",
            "key": "owasp",
            "title": "OWASP Top 10",
            "subtitle": "Web security threat distribution",
            "description": "Categorizes findings against official OWASP Top 10 2021 categories for web application security baselines.",
            "category": "Technical",
            "icon": "bug_report",
        },
    ]


@router.post("/generate")
def generate_report(
    req: ReportGenerateRequest,
    db: Session = Depends(get_db)
):
    service = ReportService(db)
    try:
        report_type = req.report_type.lower()
        fmt = req.format.lower()

        if fmt == "json":
            res_data = service.generate_json_report(req.project_id, report_type)
            return JSONResponse(content=res_data)
        elif fmt == "html":
            html_content = service.generate_html_report(req.project_id, report_type)
            return HTMLResponse(content=html_content)
        elif fmt == "pdf":
            pdf_bytes = service.generate_pdf_report(req.project_id, report_type)
            filename = f"kyptic_report_{req.project_id}_{report_type}.pdf"
            # Sanitize filename for headers
            safe_filename = re.sub(r"[^a-zA-Z0-9_\-\.]", "_", filename)
            return Response(
                content=pdf_bytes,
                media_type="application/pdf",
                headers={
                    "Content-Disposition": f'attachment; filename="{safe_filename}"'
                }
            )
        else:
            raise HTTPException(status_code=400, detail="Unsupported format. Choose 'pdf', 'html', or 'json'.")

    except ValueError as val_err:
        raise HTTPException(status_code=404, detail=str(val_err))
    except Exception as err:
        raise HTTPException(status_code=500, detail=f"Failed to generate report: {err}")


@router.get("/download")
def download_report(
    project_id: int = Query(...),
    report_type: str = Query("executive"),
    format: str = Query("pdf"),
    db: Session = Depends(get_db)
):
    """GET endpoint for downloading reports directly from browser links."""
    req = ReportGenerateRequest(project_id=project_id, report_type=report_type, format=format)
    return generate_report(req, db)
